src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK013ThreadAffinitizedMEFConstruction.cs

```diff
@@ -0,0 +1,226 @@
+        private static readonly Regex NegatableTypeOrMemberReferenceRegex = new Regex(@"^(?<negated>!)?\[(?<typeName>[^\[\]\:]+)+\](?:\:\:(?<memberName>\S+))?\s*$", RegexOptions.Singleline | RegexOptions.CultureInvariant, RegexMatchTimeout);
+        private static readonly char[] QualifiedIdentifierSeparators = new[] { '.' };
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        private ImmutableArray<TypeMatchSpec> MembersRequiringMainThread { get; set; }
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            // Register for compilation first so that we only activate the analyzer for applicable compilations
+            context.RegisterCompilationStartAction(startCompilation =>
+            {
+                this.MembersRequiringMainThread = ReadTypesAndMembers(startCompilation.Options, FileNamePatternForMembersRequiringMainThread, startCompilation.CancellationToken).ToImmutableArray();
+                INamedTypeSymbol? importingConstructorAttribute = startCompilation.Compilation.GetTypeByMetadataName(Types.ImportingConstructorAttribute.FullName)?.OriginalDefinition;
+                INamedTypeSymbol? partImportsSatisfiedNotificationInterface = startCompilation.Compilation.GetTypeByMetadataName(Types.IPartImportsSatisfiedNotification.FullName)?.OriginalDefinition;
+                INamedTypeSymbol? exportAttributeType = startCompilation.Compilation.GetTypeByMetadataName(Types.ExportAttribute.FullName)?.OriginalDefinition;
+
+                var operationKinds = ImmutableArray.Create<OperationKind>(
+                    OperationKind.MethodReference,
+                    OperationKind.InstanceReference,
+                    OperationKind.FieldReference,
+                    OperationKind.ObjectOrCollectionInitializer,
+                    OperationKind.Invocation, // Method calls
+                    OperationKind.MemberInitializer, // For static member access
+                    OperationKind.PropertyReference, // For property access
+                    OperationKind.ObjectCreation); // For object creation scenarios
+
+                startCompilation.RegisterOperationAction(
+                    Utils.DebuggableWrapper(c => this.AnalyzeOperation(c, exportAttributeType, importingConstructorAttribute, partImportsSatisfiedNotificationInterface)),
+                    operationKinds);
+            });
+        }
+
+        private static IEnumerable<TypeMatchSpec> ReadTypesAndMembers(AnalyzerOptions analyzerOptions, Regex fileNamePattern, CancellationToken cancellationToken)
```

Comment:
This is a simplification of method from `vs-threading` repo's `CommonInterest` utility class.